In [2]:
!pip install nltk
!pip install rouge-score
!pip install sacrebleu

Defaulting to user installation because normal site-packages is not writeable

[notice] A new release of pip is available: 24.2 -> 24.3.1
[notice] To update, run: python3 -m pip install --upgrade pip
Defaulting to user installation because normal site-packages is not writeable

[notice] A new release of pip is available: 24.2 -> 24.3.1
[notice] To update, run: python3 -m pip install --upgrade pip
Defaulting to user installation because normal site-packages is not writeable

[notice] A new release of pip is available: 24.2 -> 24.3.1
[notice] To update, run: python3 -m pip install --upgrade pip


In [3]:
import pandas as pd
from nltk.translate.bleu_score import sentence_bleu, SmoothingFunction
from nltk.translate.meteor_score import meteor_score
from nltk.tokenize import word_tokenize
from rouge_score import rouge_scorer
import sacrebleu
import subprocess

import nltk

# Download necessary NLTK resources
nltk.download('punkt_tab')
nltk.download('wordnet')

[nltk_data] Downloading package punkt_tab to /home/alexm/nltk_data...
[nltk_data]   Package punkt_tab is already up-to-date!
[nltk_data] Downloading package wordnet to /home/alexm/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


True

In [30]:
dataset = pd.read_csv('data/code_completion_output.csv')

In [31]:
dataset.head()

file_path  \
0          ../airflow/dags/utils/price_processing.py   
1  ../PycharmProjects/sales_telegram_bot/backend/...   
2                   ../airflow/dags/data_pipeline.py   
3          ../airflow/dags/utils/price_processing.py   
4            ../airflow/dags/utils/yolo_ocr_utils.py   

                                              prefix  \
0  import re\n\n\n# Helper function to parse pric...   
1  import json\nimport os\nfrom flask import Flas...   
2  import json\nimport ast\nimport logging\nimpor...   
3  import re\n\n\n# Helper function to parse pric...   
4  import json\nimport os\nimport tempfile\n\nimp...   

                                              middle  \
0      if len(parsed_prices) == 1:\n        retur...   
1          image_file = request.files['image']\n ...   
2      trie = Trie()\n    for word in flat_words:...   
3          return {"item_price": item_price, "ini...   
4          with open(image_path, 'rb') as image_f...   

                                              suffix  \
0  def process_billa(price_str, price_type):\n   ...   
1          torch.cuda.empty_cache()  # Clear CUDA...   
2      try:\n        detection_data = ast.literal...   
3          return {"item_price": parsed_prices[0]...   
4              return data.get('extracted_text', ...   

                                      generated_code  
0  \n\n    # Logic to compare against previous fo...  
1                                                     
2      trie = Trie()\n\n    # Run YOLO Model 2 on...  
3          parsed_prices = [parse_price(p) for p ...  
4          elif response.status_code == 200:\n   ...

## Manual Evaluation

I now want to manually evaluate the generated code. I will rate the visual 'quality' and 'similarity' of the code with scores of 0, 0.5, or 1. A score of 0 indicates no similarity, 0.5 indicates some similarities, and 1 indicates that the code is almost correct or matches exactly.

In [12]:
# Placeholder for manual scores
manual_scores = []

# Iterate over each row in the dataset
for idx, row in dataset.iterrows():
    print(f"Instance {idx + 1}")
    print("File Path:", row['file_path'])
    
    # Display Actual Code (middle) and Generated Code
    actual_code = row['middle']
    generated_code = row['generated_code']
    
    print("\nActual Code (middle):\n", actual_code)
    print("\nGenerated Code:\n", generated_code)
    
    # Prompt for manual score input
    score = input("Enter your score for similarity (0 = no similarity, 0.5 = some similarity, 1 = almost correct or exact): ")
    
    # Append score to list
    try:
        score = float(score)
        if score in {0, 0.5, 1}:
            manual_scores.append(score)
        else:
            print("Invalid input. Score must be 0, 0.5, or 1.")
            break
    except ValueError:
        print("Invalid input. Please enter a numeric score of 0, 0.5, or 1.")
        break
    
    # Divider for readability
    print("-" * 80, "\n")

# Add scores to the dataset
dataset['manual_score'] = manual_scores

# Save the dataset with scores for later analysis
dataset.to_csv('data/evaluated_code_completion_output.csv', index=False)

Instance 1
File Path: ../airflow/dags/utils/price_processing.py

Actual Code (middle):
     if len(parsed_prices) == 1:
        return {"item_price": parsed_prices[0]}

    return None


# Billa Condition


Generated Code:
 

    # Logic to compare against previous formatting (price was less than current price)
    if packaging is not None:
        new_price = packaging[: packaging.index('.')]  # Add the packaging index before replacement
        old_price = packaging[packaging.index('.') + 1:-1].strip()

        if old_price!= new_price:
            return False

    return True


# New York Condition
def process_newyork(price_str):
    return parse_price(price_str)


# Tango Condition
def process_tango(price_str):
    return parse_price


Enter your score for similarity (0 = no similarity, 0.5 = some similarity, 1 = almost correct or exact):  0


-------------------------------------------------------------------------------- 

Instance 2
File Path: ../PycharmProjects/sales_telegram_bot/backend/models_app/app.py

Actual Code (middle):
         image_file = request.files['image']
        image_file.save(temp_img_file.name)
        temp_img_path = temp_img_file.name

    try:
        # Extract text using GOT-OCR2_0 model's chat interface
        extracted_text = extract_text_from_image(temp_img_path)

        return jsonify({'extracted_text': extracted_text}), 200

    except Exception as e:
        return jsonify({'error': str(e)}), 500

    finally:
        # Clean up: Delete the temporary image file
        os.remove(temp_img_path)


Generated Code:
    


Enter your score for similarity (0 = no similarity, 0.5 = some similarity, 1 = almost correct or exact):  0


-------------------------------------------------------------------------------- 

Instance 3
File Path: ../airflow/dags/data_pipeline.py

Actual Code (middle):
     trie = Trie()
    for word in flat_words:
        trie.insert(word)

    # Parse detection_data string to dictionary


Generated Code:
     trie = Trie()

    # Run YOLO Model 2 on images and perform OCR



Enter your score for similarity (0 = no similarity, 0.5 = some similarity, 1 = almost correct or exact):  0.5


-------------------------------------------------------------------------------- 

Instance 4
File Path: ../airflow/dags/utils/price_processing.py

Actual Code (middle):
         return {"item_price": item_price, "initial_price": initial_price}

    if len(parsed_prices) == 2:
        # If the second price is commonly a "cents" part like 90 or 99, merge with the first
        if parsed_prices[1] in common_cents:
            return {"item_price": float(f"{int(parsed_prices[0])}.{int(parsed_prices[1])}")}
        else:
            return {"item_price": parsed_prices[0], "initial_price": parsed_prices[1]}

    if len(parsed_prices) == 1:


Generated Code:
         parsed_prices = [parse_price(p) for p in parsed_prices if parse_price(p) is not None]

        # Handle case where no previous price is found
        if parsed_prices and parsed_prices[0]!= parsed_prices[1]:
            return None

    # Restrict matching prices using the volume keyword and normalize by first digit
    volumes 

Enter your score for similarity (0 = no similarity, 0.5 = some similarity, 1 = almost correct or exact):  0


-------------------------------------------------------------------------------- 

Instance 5
File Path: ../airflow/dags/utils/yolo_ocr_utils.py

Actual Code (middle):
         with open(image_path, 'rb') as image_file:
            # Prepare the multipart/form-data request
            files = {'image': image_file}
            # Send JSON data separately from form-data (image)
            json_data = {'box': box}

            response = requests.post(url, files=files, data={'json': json.dumps(json_data)})

        if response.status_code == 200:
            data = response.json()


Generated Code:
         elif response.status_code == 200:
            data = response.json()

        logger.info(f"Got text from ImageBox: {data.get('extracted_text')} - {data.get('detected_boxes'):>8d}")
        return data.get('extracted_text', '')
    except Exception as e:
        logger.error(f"Exception in got_text_from_image_box: {e}")
        raise Exception(f"Exception in got_text_from_image_box: {

Enter your score for similarity (0 = no similarity, 0.5 = some similarity, 1 = almost correct or exact):  0


-------------------------------------------------------------------------------- 

Instance 6
File Path: ../PycharmProjects/sales_telegram_bot/backend/telegram_lambda_package/six.py

Actual Code (middle):
 for attr in _urllib_request_moved_attributes:
    setattr(Module_six_moves_urllib_request, attr.name, attr)


Generated Code:
 args, **kwargs)


if PY3:
    def getsetdefault(obj, default=None):
        return set(getattr(obj, _setdefault_attribute),
                   frozenset(getattr(obj, _unsetdefault_attribute)))

    def getsetnotdefault(obj, default=None):
        return set(getattr(obj, _setnotdefault_attribute),
                   getattr(obj, _unsetnotdefault_attribute))


if PY3:
    def getdelete(obj, key, default=False):
        return getattr(obj, _delelete_attribute)[key] is not None


Enter your score for similarity (0 = no similarity, 0.5 = some similarity, 1 = almost correct or exact):  0


-------------------------------------------------------------------------------- 

Instance 7
File Path: ../airflow/dags/utils/s3_dynamodb_utils.py

Actual Code (middle):
         print(f"Updated item in table {table_name} with key {key}")
    except Exception as e:


Generated Code:
     except Exception as e:



Enter your score for similarity (0 = no similarity, 0.5 = some similarity, 1 = almost correct or exact):  1


-------------------------------------------------------------------------------- 

Instance 8
File Path: ../airflow/dags/utils/correct_names.py

Actual Code (middle):
         for start in range(len(text)):
            node = self.root
            for end in range(start, len(text)):


Generated Code:
         start = 0
        end = len(text)

        while start < end:
            # We'll use the recursion limit here until we hit our target index position
            # If we're not out of bounds, then there's no solution so far
            if start >= len(text):
                break

            # If we've already reached the target index, remove it from the text and add its children
            if start == end or text[start]!= '|':
            #if start > 0 and text[start - 1] == '-' and end - 1 == start:
            #    print("Found target %


Enter your score for similarity (0 = no similarity, 0.5 = some similarity, 1 = almost correct or exact):  0


-------------------------------------------------------------------------------- 

Instance 9
File Path: ../airflow/dags/utils/price_processing.py

Actual Code (middle):
     prices = re.findall(r'\d+[.,]?\d*', price_str)


Generated Code:
     prices = re.findall(r'\d+[.,]?\d*', price_str)
    parsed_prices = [parse_price(p) for p in prices if parse_price(p) is not None]

    # Handle cena and item_member_price based on price_type



Enter your score for similarity (0 = no similarity, 0.5 = some similarity, 1 = almost correct or exact):  1


-------------------------------------------------------------------------------- 

Instance 10
File Path: ../airflow/dags/utils/correct_names.py

Actual Code (middle):
         variants.append(''.join(modified_word))

    return variants

# Trie data structure to efficiently store and search words


Generated Code:
         for position in positions:
            modified_word[position] = substitutions[modified_word[position]]

        variants.append(modified_word)
    return variants




Enter your score for similarity (0 = no similarity, 0.5 = some similarity, 1 = almost correct or exact):  1


-------------------------------------------------------------------------------- 

Instance 11
File Path: ../airflow/dags/utils/s3_dynamodb_utils.py

Actual Code (middle):
     try:


Generated Code:
     try:
        s3.upload_file(local_path, bucket_name, s3_path)
        print(f"Uploaded {local_path} to s3://{bucket_name}/")
    except Exception as e:
        print(f"Error uploading {local_path}: {e}")


# Function to delete a file from an S3 bucket
def delete_file_from_s3(filename_path, s3_path, bucket_name="salestelegrambot"):
    """
    Delete a file from an S3 bucket.
    """
    try:



Enter your score for similarity (0 = no similarity, 0.5 = some similarity, 1 = almost correct or exact):  0.5


-------------------------------------------------------------------------------- 

Instance 12
File Path: ../PycharmProjects/sales_telegram_bot/backend/telegram_lambda_package/lambda_function.py

Actual Code (middle):
     user_preferences_table = dynamodb.Table('user_preferences')
    pdf_metadata_table = dynamodb.Table('pdf_metadata')
    detected_data_table = dynamodb.Table("detected_data")


# Call the initialization function
initialize_globals()


# --------------- AWS Handling ---------------


Generated Code:
 ) if unique_shops else []


def create_shop_from_pdf_data(chat_id, data, language='en'):
    """
    Creates a new shop from the PDF data in the specified format.
    """
    # Create the new shop using the data as an empty dictionary
    shop = {}

    # Convert the data into a JSON object
    try:
        json_object = json.dumps(data, indent=4, sort_keys=True, separators=(ast.literal_eval,))
    except Exception as e:
        print("JSON exception parsing:", type(e).__n

Enter your score for similarity (0 = no similarity, 0.5 = some similarity, 1 = almost correct or exact):  0


-------------------------------------------------------------------------------- 

Instance 13
File Path: ../PycharmProjects/sales_telegram_bot/backend/telegram_lambda_package/six.py

Actual Code (middle):
 if PY3:
    _meth_func = "__func__"
    _meth_self = "__self__"

    _func_closure = "__closure__"
    _func_code = "__code__"
    _func_defaults = "__defaults__"
    _func_globals = "__globals__"
else:
    _meth_func = "im_func"
    _meth_self = "im_self"

    _func_closure = "func_closure"
    _func_code = "func_code"
    _func_defaults = "func_defaults"
    _func_globals = "func_globals"


try:


Generated Code:
 ``
    # attribute at the top so that we can safely call this function instead of
    # passing it directly to python's built-in modules.
    update_wrapper = importlib.util.update_wrapper
else:
    update_wrapper = functools.update_wrapper
    def get_child(cls, key):
        return moves.__dict__.setdefault(key, _MovedItem())


_MovedItem = TypeVar("_MovedItem", bound=

Enter your score for similarity (0 = no similarity, 0.5 = some similarity, 1 = almost correct or exact):  0


-------------------------------------------------------------------------------- 

Instance 14
File Path: ../airflow/dags/utils/pdf_utils.py

Actual Code (middle):
     file_path = os.path.join(TEMP_DIR, filename)

    # Log paths for debugging
    logging.info(f"Checking if file exists in S3 path: {s3_pdf_path}")

    logging.info(f"Downloading file from S3 path: {s3_pdf_path} to local path: {file_path}")

    try:
        download_file_from_s3(s3_pdf_path, file_path)
    except Exception as e:
        logging.error(f"Failed to download file from S3: {e}")
        raise e

    # Convert PDF into image pages
    images = convert_from_path(file_path, dpi=250, poppler_path=poppler_path)



Generated Code:
     try:
        with open(os.path.join(tempfile.gettempdir(), "upload.txt"), "w+") as f:
            f.write(str(response['Body'])))

            # Upload this PDF item to S3 in the 'pages/valid/' directory
            response = download_file_from_s3(s3_pdf_path, TEMP_DIR)
          

Enter your score for similarity (0 = no similarity, 0.5 = some similarity, 1 = almost correct or exact):  0


-------------------------------------------------------------------------------- 

Instance 15
File Path: ../airflow/dags/data_pipeline.py

Actual Code (middle):
     wait_for_others_task = ExternalTaskSensor(
        task_id='wait_for_other_pipelines',
        external_dag_id='pages_data_pipeline',  # Assuming other pipelines have the same DAG ID
        external_task_id='process_detected_items',  # Wait for this task in all other DAGs
        mode='poke',  # Can be 'reschedule' to avoid loading the database
        poke_interval=10,  # Check every 30 seconds
        trigger_rule=TriggerRule.NONE_FAILED_MIN_ONE_SUCCESS,
        dag=dag
    )

    # Task to delete endpoints at the end
    delete_endpoints_task = PythonOperator(
        task_id='delete_endpoints',
        python_callable=stop_ec2_instances,
        trigger_rule=TriggerRule.ALL_DONE,  # Only run if all upstream tasks succeed
        dag=dag
    )

    trigger_other_pipeline_task = TriggerDagRunOperator(
        task_id='

Enter your score for similarity (0 = no similarity, 0.5 = some similarity, 1 = almost correct or exact):  1


-------------------------------------------------------------------------------- 

Instance 16
File Path: ../airflow/dags/validity_check.py

Actual Code (middle):
         for file_substr in all_files:
            if file_substr in detected_image_path:
                new_valid_status = file_substr in valid_files

                # Debug: Show the new validity status for the detected item


Generated Code:
         if detected_image_path.startswith(pdf_item['image_id']):
            # The file was detected successfully
            pass
        elif detected_image_path.endswith('.pdf'):
            # The file was detected successfully but with a different file type
            pass
        else:
            # The file was detected successfully but without an extention, so ignore it
            pass

    # Add the new item to the detected_table database
    if valid_files:
        # Create the list of images we need to use when checking the validity of the detected file
        img_urls 

Enter your score for similarity (0 = no similarity, 0.5 = some similarity, 1 = almost correct or exact):  0


-------------------------------------------------------------------------------- 

Instance 17
File Path: ../airflow/dags/utils/s3_dynamodb_utils.py

Actual Code (middle):
         print(f"Updated item in table {table_name} with key {key}")
    except Exception as e:


Generated Code:
     except Exception as e:



Enter your score for similarity (0 = no similarity, 0.5 = some similarity, 1 = almost correct or exact):  1


-------------------------------------------------------------------------------- 

Instance 18
File Path: ../PycharmProjects/sales_telegram_bot/backend/telegram_lambda_package/lambda_function.py

Actual Code (middle):
     description = "Welcome to the Smart Shopping Bot! I will help you track prices, manage sale sheets, and find the best shopping paths."
    requests.post(f"{API_URL}/sendMessage", json={"chat_id": chat_id, "text": description})

    if state == "new_user":
        # Set default preferences for new users
        preferences = get_user_preferences(chat_id)
        if 'photo_group_enabled' not in preferences:
            preferences['photo_group_enabled'] = True  # Default to show photo groups
        if 'text_info_enabled' not in preferences:
            preferences['text_info_enabled'] = False  # Default to hide text info
        if 'receive_pdf_enabled' not in preferences:
            preferences['receive_pdf_enabled'] = True  # Default to send pdf after new validatio

Enter your score for similarity (0 = no similarity, 0.5 = some similarity, 1 = almost correct or exact):  0


-------------------------------------------------------------------------------- 

Instance 19
File Path: ../PycharmProjects/sales_telegram_bot/backend/telegram_lambda_package/lambda_function.py

Actual Code (middle):
                 added = add_tracked_item(chat_id, item_name)

                if added:
                    response = f"'{item_name}' saved for tracking. I will notify you when '{item_name}' has a valid sale."
                    requests.post(f"{API_URL}/sendMessage", json={"chat_id": chat_id, "text": response})
                else:
                    tracked_items = get_tracked_items(chat_id)
                    response = f"'{item_name}' is already in your tracking list. Here is your current list:\n" + "\n".join(
                        tracked_items)
                    requests.post(f"{API_URL}/sendMessage", json={"chat_id": chat_id, "text": response})
                    main_menu(chat_id)
                    save_user_state(chat_id, None)
                    re

Enter your score for similarity (0 = no similarity, 0.5 = some similarity, 1 = almost correct or exact):  0


-------------------------------------------------------------------------------- 

Instance 20
File Path: ../airflow/dags/utils/pdf_utils.py

Actual Code (middle):
         raise Exception("Filename or Shop Name missing!")

    # Fetch metadata from DynamoDB
    response = get_pdf_item_from_dynamodb(filename, shop_name)
    file_entry = response.get('Item')

    if not file_entry:
        raise Exception(f"File {filename} not found in DynamoDB")

    # Check if the pages already exist in S3


Generated Code:
         raise ValueError("The input argument is required")




Enter your score for similarity (0 = no similarity, 0.5 = some similarity, 1 = almost correct or exact):  0.5


-------------------------------------------------------------------------------- 

Instance 21
File Path: ../PycharmProjects/sales_telegram_bot/backend/models_app/app.py

Actual Code (middle):
         image_file = request.files['image']
        image_file.save(temp_img_file.name)
        temp_img_path = temp_img_file.name

    try:
        # Read the image using OpenCV
        img = cv2.imread(temp_img_path)

        # Select model based on query param, default to model1


Generated Code:
     img = cv2.imread(os.path.join(request.files["image"].filename)))

    try:
        # Perform the actual prediction and return its classification label and confidence score



Enter your score for similarity (0 = no similarity, 0.5 = some similarity, 1 = almost correct or exact):  0.5


-------------------------------------------------------------------------------- 

Instance 22
File Path: ../airflow/dags/validity_check.py

Actual Code (middle):
         included_shops = [shop for shop in all_shops if shop not in excluded_shops]

        # Iterate over all included shops and add users to corresponding shop keys
        for shop in included_shops:
            # Add to "included_shops" column (all users with included shops)
            if shop not in shop_user_map["included_shops"]:
                shop_user_map["included_shops"][shop] = []
            shop_user_map["included_shops"][shop].append(chat_id)

            # Add to "included_shops_and_receive_pdf" column (users with included shops AND receive_pdf_enabled=True)
            if receive_pdf_enabled:
                if shop not in shop_user_map["included_shops_and_receive_pdf"]:
                    shop_user_map["included_shops_and_receive_pdf"][shop] = []
                shop_user_map["included_shops_and_receiv

Enter your score for similarity (0 = no similarity, 0.5 = some similarity, 1 = almost correct or exact):  0.5


-------------------------------------------------------------------------------- 

Instance 23
File Path: ../airflow/dags/utils/yolo_ocr_utils.py

Actual Code (middle):
             files = {'image': image_file}
            # Send JSON data separately from form-data (image)
            json_data = {'box': box}

            response = requests.post(url, files=files, data={'json': json.dumps(json_data)})

        if response.status_code == 200:
            data = response.json()
            return data.get('extracted_text', '')
        else:
            raise Exception(f"Error in extract_text_from_image_box: {response.status_code} - {response.text}")
    except Exception as e:
        logger.error(f"Exception in got_text_from_image_box: {e}")
        raise Exception(f"Exception in extract_text_from_image_box: {e}")




Generated Code:
             formdata = {"image": image_file, "box": box}

            headers = {'Content-Type': 'application/octet-stream'}

            resp = requests.

Enter your score for similarity (0 = no similarity, 0.5 = some similarity, 1 = almost correct or exact):  1


-------------------------------------------------------------------------------- 

Instance 24
File Path: ../PycharmProjects/sales_telegram_bot/backend/sales_telegram_bot_admin_backend/app.py

Actual Code (middle):
         if 'Authorization' in request.headers:
            try:
                # Extract token from "Bearer <token>"


Generated Code:
             try:
            # Try to authenticate
            auth = (current_user, token)
            r = requests.post(AIRFLOW_URL, headers={"Content-Type": "application/x-www-form-urlencoded"),
                             params={'grant_type':'password',
                                    'username': current_user,
                                    'password': <PASSWORD>,
                                  'scope': 'all'}, data=auth)
            
            # Raise exception if we have any errors
            if r.ok:
           


Enter your score for similarity (0 = no similarity, 0.5 = some similarity, 1 = almost correct or exact):  0.5


-------------------------------------------------------------------------------- 

Instance 25
File Path: ../PycharmProjects/sales_telegram_bot/backend/sales_telegram_bot_admin_backend/app.py

Actual Code (middle):
         print(f"Provided password type: {type(password)}")        # Should be str

        # Check if either value is None or invalid
        if not stored_password or not isinstance(stored_password, str):
            return jsonify({"error": "Stored password is invalid"}), 500

        if not check_password(stored_password, password):
            return jsonify({"error": "Invalid password"}), 401

        # Generate JWT token upon successful login


Generated Code:
         if type(stored_password)!= type(default_password):
            # The stored password type does not match the expected format
            print(f"The stored password type: {type(stored_password)} is different than the expected format: {type(default_password)})")  # Debug: Log different stored passwords



Enter your score for similarity (0 = no similarity, 0.5 = some similarity, 1 = almost correct or exact):  0.5


-------------------------------------------------------------------------------- 

Instance 26
File Path: ../PycharmProjects/sales_telegram_bot/backend/sales_telegram_bot_admin_backend/app.py

Actual Code (middle):
     try:
        # Load PDF data
        pdf_data = load_pdf_data()

        # Find the entry in the DynamoDB table
        file_entry = next((entry for entry in pdf_data if entry['filename'] == filename), None)

        if not file_entry:
            return jsonify({"error": "File not found"}), 404

        # Delete the file from S3
        s3.delete_object(Bucket=BUCKET_NAME, Key=f'pdfs/{filename}')

        # Remove the entry from DynamoDB
        table = dynamodb.Table(PDF_TABLE_NAME)
        table.delete_item(
            Key={
                'filename': filename,
                'shop_name': file_entry['shop_name']
            }
        )

        # Remove the file entry from the local pdf_data if applicable
        pdf_data = [entry for entry in pdf_data if entry['f

Enter your score for similarity (0 = no similarity, 0.5 = some similarity, 1 = almost correct or exact):  0


-------------------------------------------------------------------------------- 

Instance 27
File Path: ../airflow/dags/utils/price_processing.py

Actual Code (middle):
     clean_str = price_str.replace("€", "").strip()

    # Find all the price values in the string
    prices = re.findall(r'\d+[.,]?\d*', clean_str)
    parsed_prices = [parse_price(p) for p in prices if parse_price(p) is not None]

    # Ensure prices are sorted correctly (sale price is less than initial price)
    if len(parsed_prices) == 2:
        sale_price = min(parsed_prices)
        initial_price = max(parsed_prices)
        return {"item_price": sale_price, "initial_price": initial_price}

    # If we only have one price, return it as the item price
    elif len(parsed_prices) == 1:


Generated Code:
     if '%' in price_str or '(' in price_str:
        return None

    # Handle cases like "19 90 25.90 2" by removing the "£" sign with decimals
    price_str = price_str.replace('£', '')

    # Extract all num

Enter your score for similarity (0 = no similarity, 0.5 = some similarity, 1 = almost correct or exact):  1


-------------------------------------------------------------------------------- 

Instance 28
File Path: ../airflow/dags/utils/yolo_ocr_utils.py

Actual Code (middle):
             response = requests.post(url, files=files)

        if response.status_code == 200:
            data = response.json()
            return data.get('extracted_text', '')
        else:
            raise Exception(f"Error in extract_text_from_image: {response.status_code} - {response.text}")
    except Exception as e:
        raise Exception(f"Exception in extract_text_from_image: {e}")




Generated Code:
         response = requests.post(url, files=files)

        if response.status_code == 200:
            content = response.content
        elif response.status_code == 404:
            return None  # No such page exists
        else:
            content = response.text

        return content
    except Exception as e:
        logger.error(f"Exception in got_text_from_image: {e}")
        return None




Enter your score for similarity (0 = no similarity, 0.5 = some similarity, 1 = almost correct or exact):  1


-------------------------------------------------------------------------------- 

Instance 29
File Path: ../airflow/dags/utils/pdf_utils.py

Actual Code (middle):
     response = get_pdf_item_from_dynamodb(filename, shop_name)
    file_entry = response.get('Item')

    if not file_entry:
        raise Exception(f"File {filename} not found in DynamoDB")

    # Check if the pages already exist in S3
    page_s3_paths = []  # Store full S3 paths for pages


Generated Code:
     dynamodb_table = boto3.resource('dynamodb', region_name='us-east-1')
    table = dynamodb_table['shop']

    response = table.query(TableName=shop_name)['Items']

    items = []
    for item in response:
        items.append({"id": item['Id'], "title": item["Title"], "link": item["Link"]})

    page_list = [items[0]]
    page_s3_paths = []




Enter your score for similarity (0 = no similarity, 0.5 = some similarity, 1 = almost correct or exact):  1


-------------------------------------------------------------------------------- 

Instance 30
File Path: ../airflow/dags/utils/s3_dynamodb_utils.py

Actual Code (middle):
     try:
        table = table or dynamodb.Table(table_name)
        table.update_item(
            Key=key,
            UpdateExpression=update_expression,
            ExpressionAttributeValues=expression_attribute_values
        )


Generated Code:
     try:
        table = table or dynamodb.Table(table_name)
        table[key] = update_expression
        for attribute_value in expression_attribute_values:
            table["Attributes"][attribute_value['AttributeName']] = attribute_value['Value']
        print(f"Updated items in table {table_name} with key {key}")
    except Exception as e:
        print(f"Error updating items in DynamoDB: {e}")

# Function to delete specific fields of an item in DynamoDB
def delete_item_in_dynamodb(table_name, key, table=


Enter your score for similarity (0 = no similarity, 0.5 = some similarity, 1 = almost correct or exact):  0.5


-------------------------------------------------------------------------------- 

Instance 31
File Path: ../airflow/dags/validity_check.py

Actual Code (middle):
     detected_items = detected_response.get('Items', [])

    print(f"Debug: Detected items retrieved from DynamoDB: {detected_items}")

    # Process each detected item
    for detected_item in detected_items:
        detected_image_path = detected_item['image_id']

        # Debug: Show the current detected image path
        print(f"Debug: Processing detected item with image_id: {detected_image_path}")

        # Check if any file name (without .pdf) is a substring of the image_id
        for file_substr in all_files:
            if file_substr in detected_image_path:


Generated Code:
     detected_images = []

    # Loop through every item in the table to find all files that are currently being processed
    for item in detected_response['Items':
        # Find all images for this image id
        if item.get('img_id'):


Enter your score for similarity (0 = no similarity, 0.5 = some similarity, 1 = almost correct or exact):  0


-------------------------------------------------------------------------------- 

Instance 32
File Path: ../PycharmProjects/sales_telegram_bot/backend/sales_telegram_bot_admin_backend/app.py

Actual Code (middle):
         response = users_table.get_item(Key={'username': username})
        user = response.get('Item')

        if not user:
            return jsonify({"error": "Invalid username"}), 401

        stored_password = user.get('password')

        # Debug: Print the types of variables
        print(f"Stored password type: {type(stored_password)}")  # Should be str
        print(f"Provided password type: {type(password)}")        # Should be str

        # Check if either value is None or invalid
        if not stored_password or not isinstance(stored_password, str):
            return jsonify({"error": "Stored password is invalid"}), 500



Generated Code:
         stored_password = str(password)
        database_credentials = USERS_TABLE_.query(f"username=='{username}'").fir

Enter your score for similarity (0 = no similarity, 0.5 = some similarity, 1 = almost correct or exact):  0


-------------------------------------------------------------------------------- 

Instance 33
File Path: ../PycharmProjects/sales_telegram_bot/backend/telegram_lambda_package/six.py

Actual Code (middle):
     if isinstance(attr, MovedModule):
        _importer._add_module(attr, "moves." + attr.name)
del attr

_MovedItems._moved_attributes = _moved_attributes

moves = _MovedItems(__name__ + ".moves")


Generated Code:
 ("moves.urllib_error")
    request = _importer._get_module("moves.urllib_request")
    response = _importer._get_module("moves.urllib_response")
    delattr(_MovedItems, attr.name)


class _MovedModules(_LazyModule):

    """Lazy loading of moved modules from an iterable"""
    __path__ = []  # mark as package


_moved_attributes = [
    MovedAttribute("isatty", "bool", "True", "False"),
    MovedAttribute("isclosed", "bool", "True", "False"),
   


Enter your score for similarity (0 = no similarity, 0.5 = some similarity, 1 = almost correct or exact):  0


-------------------------------------------------------------------------------- 

Instance 34
File Path: ../PycharmProjects/sales_telegram_bot/backend/models_app/app.py

Actual Code (middle):
         chosen_model = request.args.get('model', 'model1')
        if chosen_model == 'model1':
            model = model1
        else:
            model = model2

        # Run YOLO detection on the image
        detected_img, results = predict_and_detect(model, img, conf=0.5)

        # Convert image to base64
        base64_image = image_to_base64(detected_img)

        # Convert results into JSON format
        result_data = []
        for result in results:
            for box in result.boxes:
                result_data.append({
                    'class': result.names[int(box.cls[0])],
                    'confidence': box.conf[0].item(),
                    'box': [int(box.xyxy[0][0]), int(box.xyxy[0][1]), int(box.xyxy[0][2]), int(box.xyxy[0][3])]
                })

        return jso

Enter your score for similarity (0 = no similarity, 0.5 = some similarity, 1 = almost correct or exact):  0


-------------------------------------------------------------------------------- 

Instance 35
File Path: ../PycharmProjects/sales_telegram_bot/backend/sales_telegram_bot_admin_backend/app.py

Actual Code (middle):
     payload = {
        'username': username,
        'exp': datetime.utcnow() + timedelta(hours=1)  # Token expires in 1 hour
    }
    token = jwt.encode(payload, SECRET_KEY, algorithm='HS256')


Generated Code:
     token = jwt.encode({
        'exp': datetime.utcnow() + timedelta(days=1)).decode('utf-8'),
      'sub': hashlib.sha256(bytes(username, encoding="UTF-8")).hexdigest(),
        'type':'secret',
        'username': username.lower()
    }, algorithm='HS256').decode('utf-8')




Enter your score for similarity (0 = no similarity, 0.5 = some similarity, 1 = almost correct or exact):  0.5


-------------------------------------------------------------------------------- 

Instance 36
File Path: ../airflow/dags/validity_check.py

Actual Code (middle):
     for pdf_item in pdf_items:
        # Remove .pdf from filename for comparison with detected items
        pdf_name_without_ext = pdf_item['filename'].replace('.pdf', '')

        valid_from = datetime.strptime(pdf_item['valid_from'], '%Y-%m-%d').date()
        valid_to = datetime.strptime(pdf_item['valid_to'], '%Y-%m-%d').date()

        # Determine the current validity status
        current_valid_status = pdf_item.get('valid', None)

        # Check if the file should be valid or invalid based on today's date


Generated Code:
     for pdf_item in pdf_items:
        # Ignore non-existing files
        if '.pdf' in pdf_item['filename':
            continue

        # Get all the photos in the pdf item
        photo_filenames = ['photo/' + str(i) + '.jpg' for i in range(64)}

        # Filter out photos that are part of 

Enter your score for similarity (0 = no similarity, 0.5 = some similarity, 1 = almost correct or exact):  0.5


-------------------------------------------------------------------------------- 

Instance 37
File Path: ../PycharmProjects/sales_telegram_bot/backend/sales_telegram_bot_admin_backend/app.py

Actual Code (middle):
         stored_password = user.get('password')

        # Debug: Print the types of variables
        print(f"Stored password type: {type(stored_password)}")  # Should be str
        print(f"Provided password type: {type(password)}")        # Should be str

        # Check if either value is None or invalid
        if not stored_password or not isinstance(stored_password, str):
            return jsonify({"error": "Stored password is invalid"}), 500



Generated Code:
     except NoCredentialsError:
        return jsonify({"error": 'Username does not exist'}), 401
    except Exception as e:
        app.logger.error(f'Error generating access token: {e}')
        return jsonify({"error": f'Error generating access token: {e}'})


@app.route('/', methods=["GET"]))
@token_requir

Enter your score for similarity (0 = no similarity, 0.5 = some similarity, 1 = almost correct or exact):  0


-------------------------------------------------------------------------------- 

Instance 38
File Path: ../PycharmProjects/sales_telegram_bot/backend/telegram_lambda_package/six.py

Actual Code (middle):
 _urllib_parse_moved_attributes = [
    MovedAttribute("ParseResult", "urlparse", "urllib.parse"),
    MovedAttribute("SplitResult", "urlparse", "urllib.parse"),
    MovedAttribute("parse_qs", "urlparse", "urllib.parse"),
    MovedAttribute("parse_qsl", "urlparse", "urllib.parse"),
    MovedAttribute("urldefrag", "urlparse", "urllib.parse"),
    MovedAttribute("urljoin", "urlparse", "urllib.parse"),
    MovedAttribute("urlparse", "urlparse", "urllib.parse"),
    MovedAttribute("urlsplit", "urlparse", "urllib.parse"),
    MovedAttribute("urlunparse", "urlparse", "urllib.parse"),
    MovedAttribute("urlunsplit", "urlparse", "urllib.parse"),
    MovedAttribute("quote", "urllib", "urllib.parse"),
    MovedAttribute("quote_plus", "urllib", "urllib.parse"),
    MovedAttribute("unquote", "u

Enter your score for similarity (0 = no similarity, 0.5 = some similarity, 1 = almost correct or exact):  0


-------------------------------------------------------------------------------- 

Instance 39
File Path: ../PycharmProjects/sales_telegram_bot/backend/telegram_lambda_package/six.py

Actual Code (middle):
     response = _importer._get_module("moves.urllib_response")


Generated Code:
 
            elif not isinstance(end, str):
            raise TypeError('end must be None or a string')
        argspec = inspect.getfullargspec(type(value), (), dict(vars()))
        defaults = kwargs.copy()
        if len(argspec.args) == 0:
            if hasattr(value, "__repr__"):
                argspec = inspect.getfullargspec(value)
            default_arg = getattr(value, "_default", None)

            if default_arg is not None:
                argspec.defaults += ("--%s" % (str(default_arg))))
           


Enter your score for similarity (0 = no similarity, 0.5 = some similarity, 1 = almost correct or exact):  0


-------------------------------------------------------------------------------- 

Instance 40
File Path: ../airflow/dags/utils/yolo_ocr_utils.py

Actual Code (middle):
         with open(image_path, 'rb') as image_file:
            files = {'image': image_file}
            response = requests.post(url, files=files)

        if response.status_code == 200:
            data = response.json()


Generated Code:
         logger.info("Getting text from image: " + image_path)
        with open(image_path, 'r') as image_file:
            data = image_file.read()  # Read image content as bytes
            logger.debug(f"Got image contents:\n{data}\n")
    except Exception as e:
        logger.error(f"Exception in got_text_from_image: {e}")
        raise Exception(f"Exception in got_text_from_image: {e}")
    finally:
        return data


def got_text_from_image_box


Enter your score for similarity (0 = no similarity, 0.5 = some similarity, 1 = almost correct or exact):  0.5


-------------------------------------------------------------------------------- 

Instance 41
File Path: ../PycharmProjects/sales_telegram_bot/backend/telegram_lambda_package/lambda_function.py

Actual Code (middle):
             photo_group_enabled = is_photo_group_enabled(chat_id)
            text_info_enabled = is_text_info_enabled(chat_id)

            # Iterate over each shop and its associated items
            for shop_name, shop_items in items_by_shop.items():
                # Prepare the initial response message


Generated Code:
 
                      json={"chat_id": chat_id, "text": f"Shop '{text}' included from tracking.'})
        settings_menu(chat_id)
        save_user_state(chat_id, 'in_settings')


def handle_exclude_shop_state(chat_id, text):
    if text == "⬅️ Back to settings":
        settings_menu(chat_id)
        save_user_state(chat_id, 'in_settings')
    else:
        exclude_shop(chat_id, text)
        requests.post(f"{API_


Enter your score for similarity (0 = no similarity, 0.5 = some similarity, 1 = almost correct or exact):  0


-------------------------------------------------------------------------------- 

Instance 42
File Path: ../airflow/dags/utils/s3_dynamodb_utils.py

Actual Code (middle):
         table = table or dynamodb.Table(table_name)
        table.update_item(
            Key=key,
            UpdateExpression=update_expression,
            ExpressionAttributeValues=expression_attribute_values
        )
        print(f"Updated item in table {table_name} with key {key}")
    except Exception as e:


Generated Code:
         table = table or dynamodb.Table(table_name)

        if not isinstance(key, str):
            raise TypeError("The value for key must be a string")

        if update_expression is None:
            return table.update_item(**{
                key: [value] if len(value) > 0 else []
                for (key, value) in zip(key, expression_attribute_values)]))
        
        elif update_expression == "ADD":
            attribute_values = list(map(lambda v: v[1:], expression_att

Enter your score for similarity (0 = no similarity, 0.5 = some similarity, 1 = almost correct or exact):  0.5


-------------------------------------------------------------------------------- 

Instance 43
File Path: ../PycharmProjects/sales_telegram_bot/backend/models_app/app.py

Actual Code (middle):
     with tempfile.NamedTemporaryFile(suffix=".png", delete=False) as temp_img_file:
        # Load the image file from the request
        image_file = request.files['image']
        image_file.save(temp_img_file.name)
        temp_img_path = temp_img_file.name



Generated Code:
     temp_img_file = None
    try:
    


Enter your score for similarity (0 = no similarity, 0.5 = some similarity, 1 = almost correct or exact):  0


-------------------------------------------------------------------------------- 

Instance 44
File Path: ../airflow/dags/utils/pdf_utils.py

Actual Code (middle):
     s3_pdf_path = f'{PDF_S3_PATH}/{filename}'

    # Download the PDF from S3 to a temporary location
    file_path = os.path.join(TEMP_DIR, filename)

    # Log paths for debugging
    logging.info(f"Checking if file exists in S3 path: {s3_pdf_path}")



Generated Code:
     file_path = f'{PDF_S3_PATH}/{base_filename}'




Enter your score for similarity (0 = no similarity, 0.5 = some similarity, 1 = almost correct or exact):  1


-------------------------------------------------------------------------------- 

Instance 45
File Path: ../airflow/dags/utils/correct_names.py

Actual Code (middle):
         variants = generate_il1rjeo_combinations(word)


Generated Code:
         try:
            self.insert_into_trie(word)
        except Exception as e:
            print("Error occurred while inserting the word")
            raise e


# Implementation of the Fuzzy Matcher algorithm described here: https://en.wikipedia.org/wiki/fuzzy_matcher
# Adapted from http://www.programmers.co.uk/learn/2014/06/fuzzy-matching.html
class FuzzyMatcher():
    def __init__(self):
        pass

    def match(self, word, document):
        raise NotImplementedError("Abstract method should implement this.")


#


Enter your score for similarity (0 = no similarity, 0.5 = some similarity, 1 = almost correct or exact):  0


-------------------------------------------------------------------------------- 

Instance 46
File Path: ../PycharmProjects/sales_telegram_bot/backend/models_app/app.py

Actual Code (middle):
         temp_img_path = temp_img_file.name

    try:
        # Read the image using OpenCV
        img = cv2.imread(temp_img_path)

        # Select model based on query param, default to model1
        chosen_model = request.args.get('model', 'model1')
        if chosen_model == 'model1':
            model = model1
        else:
            model = model2

        # Run YOLO detection on the image
        detected_img, results = predict_and_detect(model, img, conf=0.5)

        # Convert image to base64
        base64_image = image_to_base64(detected_img)

        # Convert results into JSON format
        result_data = []


Generated Code:
     # Load the image as PIL Image object
    image = cv2.imread(image_file)

    # Detect all objects within the current image
    results = []
    # Itera

Enter your score for similarity (0 = no similarity, 0.5 = some similarity, 1 = almost correct or exact):  0


-------------------------------------------------------------------------------- 

Instance 47
File Path: ../airflow/dags/validity_check.py

Actual Code (middle):
     invalid_files = [file.replace('.pdf', '') for file in invalid_files]

    # Combine valid and invalid files
    all_files = valid_files + invalid_files
    print(f"Debug: all_files to be checked (without .pdf): {all_files}")

    valid_changed_items = []

    detected_response = detected_table.scan()
    detected_items = detected_response.get('Items', [])

    print(f"Debug: Detected items retrieved from DynamoDB: {detected_items}")

    # Process each detected item


Generated Code:
     invalid_files = status_change.get('invalid', [])

    # Define the variables needed for the API calls
    processed_files = list(set(valid_files + invalid_files))

    # Create a dictionary to store processed_files
    processed_keys = {'processed_files': [], 'processed_keys': {}}

    # Loop through all files to find valid and invalid 

Enter your score for similarity (0 = no similarity, 0.5 = some similarity, 1 = almost correct or exact):  0.5


-------------------------------------------------------------------------------- 

Instance 48
File Path: ../PycharmProjects/sales_telegram_bot/backend/models_app/app.py

Actual Code (middle):
     with tempfile.NamedTemporaryFile(suffix=".png", delete=False) as temp_img_file:
        image_file = request.files['image']
        image_file.save(temp_img_file.name)
        temp_img_path = temp_img_file.name

    try:
        # Extract text using GOT-OCR2_0 model's chat interface
        extracted_text = extract_text_from_image(temp_img_path)

        return jsonify({'extracted_text': extracted_text}), 200

    except Exception as e:
        return jsonify({'error': str(e)}), 500

    finally:
        # Clean up: Delete the temporary image file
        os.remove(temp_img_path)


Generated Code:
     with tempfile.NamedTemporaryFile(suffix=".png", delete=False) as temp_img_file:
        # Load the image file from the request
        image_file = request.files['image']
        image_file.sa

Enter your score for similarity (0 = no similarity, 0.5 = some similarity, 1 = almost correct or exact):  0.5


-------------------------------------------------------------------------------- 

Instance 49
File Path: ../airflow/dags/utils/pdf_utils.py

Actual Code (middle):
         page_path = os.path.join(TEMP_DIR, page_filename)

        # Save the image locally
        image.save(page_path, 'PNG')

        # Upload each page to S3 in the 'pages/valid/' directory
        s3_page_path = f'{PAGES_S3_PATH}/{page_filename}'
        upload_file_to_s3(page_path, s3_page_path)

        # Add full S3 path of the page to the list
        page_s3_paths.append(s3_page_path)

    # Return the list of full S3 paths for the uploaded pages


Generated Code:
         # Upload the page to S3
        file_upload = upload_file_to_s3(
            file_path=file_path, image_data=image['Contents'][0], shop_name=shop_name,
            page_id='{}'.format(i),
            content="{} {}".format(image["Content"]["PageMetadata"], image["Content"]["Body"])))

        # Save page's PDF to S3
        page_s3_paths.append

Enter your score for similarity (0 = no similarity, 0.5 = some similarity, 1 = almost correct or exact):  1


-------------------------------------------------------------------------------- 

Instance 50
File Path: ../airflow/dags/utils/correct_names.py

Actual Code (middle):
         variants = generate_il1rjeo_combinations(word)


Generated Code:
         """
        Inserts a word into the Trie. If it's already present in the Trie, then replaces this word with another one.
        Words should be lower cased, separated by commas, and no spaces or punctuation.

        :param word: The word to add
        :type word: str
        :return: None
        :rtype: None
        """
        if not word:
            raise ValueError("Word cannot be empty.")
        if word[-2:] == ",":
            word += ","
        try:
            word = int(word)
        except TypeError:
            pass

        if word < 0:
            raise


Enter your score for similarity (0 = no similarity, 0.5 = some similarity, 1 = almost correct or exact):  0


-------------------------------------------------------------------------------- 



In [24]:
eval_df = pd.read_csv('evaluated_code_completion_output.csv')

In [25]:
eval_df.head()

file_path  \
0          ../airflow/dags/utils/price_processing.py   
1  ../PycharmProjects/sales_telegram_bot/backend/...   
2                   ../airflow/dags/data_pipeline.py   
3          ../airflow/dags/utils/price_processing.py   
4            ../airflow/dags/utils/yolo_ocr_utils.py   

                                              prefix  \
0  import re\n\n\n# Helper function to parse pric...   
1  import json\nimport os\nfrom flask import Flas...   
2  import json\nimport ast\nimport logging\nimpor...   
3  import re\n\n\n# Helper function to parse pric...   
4  import json\nimport os\nimport tempfile\n\nimp...   

                                              middle  \
0      if len(parsed_prices) == 1:\n        retur...   
1          image_file = request.files['image']\n ...   
2      trie = Trie()\n    for word in flat_words:...   
3          return {"item_price": item_price, "ini...   
4          with open(image_path, 'rb') as image_f...   

                                              suffix  \
0  def process_billa(price_str, price_type):\n   ...   
1          torch.cuda.empty_cache()  # Clear CUDA...   
2      try:\n        detection_data = ast.literal...   
3          return {"item_price": parsed_prices[0]...   
4              return data.get('extracted_text', ...   

                                      generated_code  manual_score  
0  \n\n    # Logic to compare against previous fo...           0.0  
1                                                              0.0  
2      trie = Trie()\n\n    # Run YOLO Model 2 on...           0.5  
3          parsed_prices = [parse_price(p) for p ...           0.0  
4          elif response.status_code == 200:\n   ...           0.0

## Automatic evaluation

The script under evaluates the generated code completions using six metrics that capture syntactic and semantic similarity between generated and reference code. The following metrics are included:
* Exact Match: Checks if the generated code exactly matches the reference code, providing a binary evaluation of precision.
* BLEU Score: Measures the overlap of n-grams, with higher scores indicating better exact matching of phrases.
* ROUGE-L: Focuses on the longest common subsequence, rewarding sequences that capture similar phrases or structural patterns.
* METEOR: Considers synonyms and stemming, enhancing evaluation for semantically similar but syntactically different phrases.
* ChrF: A character-level metric that rewards structural similarity, ideal for comparing code with potential variations in variable names.

In [26]:
# Initialize lists to store metric results
exact_matches = []
bleu_scores = []
rouge_l_scores = []
meteor_scores = []
chrf_scores = []

# Calculate metrics for each row, comparing only the middle part for all except ChrF
for index, row in eval_df.iterrows():
    # Reference and generated middle parts
    reference_middle = row['middle'].strip()
    generated_code = row['generated_code'].strip()
    
    # Find the generated middle part by locating the first instance of suffix
    suffix = row['suffix'].strip()
    if suffix in generated_code:
        generated_middle = generated_code.split(suffix)[0].strip()
    else:
        generated_middle = generated_code

    # Exact Match
    exact_match = 1.0 if reference_middle == generated_middle else 0.0
    exact_matches.append(exact_match)

    # BLEU Score with smoothing
    bleu_score = sentence_bleu([reference_middle.split()], generated_middle.split(), smoothing_function=SmoothingFunction().method1)
    bleu_scores.append(bleu_score)

    # ROUGE-L Score
    rouge = rouge_scorer.RougeScorer(['rougeL'], use_stemmer=True)
    rouge_l = rouge.score(reference_middle, generated_middle)['rougeL'].fmeasure
    rouge_l_scores.append(rouge_l)

    # METEOR Score
    meteor = meteor_score([word_tokenize(reference_middle)], word_tokenize(generated_middle))
    meteor_scores.append(meteor)

    # ChrF Score using full reference and generated code
    reference_full = f"{row['prefix'].strip()} {reference_middle} {suffix}"
    generated_full = f"{row['prefix'].strip()} {generated_middle} {suffix}"
    chrf_score = sacrebleu.corpus_chrf([reference_full], [generated_full])
    chrf_scores.append(chrf_score.score)

# Add metric columns to the DataFrame
eval_df['Exact Match'] = exact_matches
eval_df['BLEU'] = bleu_scores
eval_df['ROUGE-L'] = rouge_l_scores
eval_df['METEOR'] = meteor_scores
eval_df['ChrF'] = chrf_scores

# Save updated DataFrame to a new CSV
output_path = 'data/evaluated_code_completion_output_with_metrics.csv'
eval_df.to_csv(output_path, index=False)

## Manual and Automatic evaluation comparison

In [27]:
full_eval_df = pd.read_csv('data/evaluated_code_completion_output_with_metrics.csv')

In [28]:
full_eval_df.head(50)

file_path  \
0           ../airflow/dags/utils/price_processing.py   
1   ../PycharmProjects/sales_telegram_bot/backend/...   
2                    ../airflow/dags/data_pipeline.py   
3           ../airflow/dags/utils/price_processing.py   
4             ../airflow/dags/utils/yolo_ocr_utils.py   
5   ../PycharmProjects/sales_telegram_bot/backend/...   
6          ../airflow/dags/utils/s3_dynamodb_utils.py   
7              ../airflow/dags/utils/correct_names.py   
8           ../airflow/dags/utils/price_processing.py   
9              ../airflow/dags/utils/correct_names.py   
10         ../airflow/dags/utils/s3_dynamodb_utils.py   
11  ../PycharmProjects/sales_telegram_bot/backend/...   
12  ../PycharmProjects/sales_telegram_bot/backend/...   
13                 ../airflow/dags/utils/pdf_utils.py   
14                   ../airflow/dags/data_pipeline.py   
15                  ../airflow/dags/validity_check.py   
16         ../airflow/dags/utils/s3_dynamodb_utils.py   
17  ../PycharmProjects/sales_telegram_bot/backend/...   
18  ../PycharmProjects/sales_telegram_bot/backend/...   
19                 ../airflow/dags/utils/pdf_utils.py   
20  ../PycharmProjects/sales_telegram_bot/backend/...   
21                  ../airflow/dags/validity_check.py   
22            ../airflow/dags/utils/yolo_ocr_utils.py   
23  ../PycharmProjects/sales_telegram_bot/backend/...   
24  ../PycharmProjects/sales_telegram_bot/backend/...   
25  ../PycharmProjects/sales_telegram_bot/backend/...   
26          ../airflow/dags/utils/price_processing.py   
27            ../airflow/dags/utils/yolo_ocr_utils.py   
28                 ../airflow/dags/utils/pdf_utils.py   
29         ../airflow/dags/utils/s3_dynamodb_utils.py   
30                  ../airflow/dags/validity_check.py   
31  ../PycharmProjects/sales_telegram_bot/backend/...   
32  ../PycharmProjects/sales_telegram_bot/backend/...   
33  ../PycharmProjects/sales_telegram_bot/backend/...   
34  ../PycharmProjects/sales_telegram_bot/backend/...   
35                  ../airflow/dags/validity_check.py   
36  ../PycharmProjects/sales_telegram_bot/backend/...   
37  ../PycharmProjects/sales_telegram_bot/backend/...   
38  ../PycharmProjects/sales_telegram_bot/backend/...   
39            ../airflow/dags/utils/yolo_ocr_utils.py   
40  ../PycharmProjects/sales_telegram_bot/backend/...   
41         ../airflow/dags/utils/s3_dynamodb_utils.py   
42  ../PycharmProjects/sales_telegram_bot/backend/...   
43                 ../airflow/dags/utils/pdf_utils.py   
44             ../airflow/dags/utils/correct_names.py   
45  ../PycharmProjects/sales_telegram_bot/backend/...   
46                  ../airflow/dags/validity_check.py   
47  ../PycharmProjects/sales_telegram_bot/backend/...   
48                 ../airflow/dags/utils/pdf_utils.py   
49             ../airflow/dags/utils/correct_names.py   

                                               prefix  \
0   import re\n\n\n# Helper function to parse pric...   
1   import json\nimport os\nfrom flask import Flas...   
2   import json\nimport ast\nimport logging\nimpor...   
3   import re\n\n\n# Helper function to parse pric...   
4   import json\nimport os\nimport tempfile\n\nimp...   
5   # Copyright (c) 2010-2020 Benjamin Peterson\n#...   
6   import boto3\n\n# Initialize AWS S3 and Dynamo...   
7   import re\nimport hunspell\nimport itertools\n...   
8   import re\n\n\n# Helper function to parse pric...   
9   import re\nimport hunspell\nimport itertools\n...   
10  import boto3\n\n# Initialize AWS S3 and Dynamo...   
11  import ast\nimport copy\nimport re\nimport tem...   
12  # Copyright (c) 2010-2020 Benjamin Peterson\n#...   
13  import os\nfrom pdf2image import convert_from_...   
14  import json\nimport ast\nimport logging\nimpor...   
15  import itertools\n\nfrom airflow.sensors.exter...   
16  import boto3\n\n# Initialize AWS S3 and Dynamo...   
17  import ast\nimport copy\nimport re\nimport tem...   
18  import ast\nimport copy\nimport re\nimport tem...

In [29]:
# Select relevant columns for correlation analysis
metrics = ['Exact Match', 'BLEU', 'ROUGE-L', 'METEOR', 'ChrF']
correlations = {}

# Calculate Spearman and Kendall correlations for each metric with manual score
for metric in metrics:
    spearman_corr = eval_df['manual_score'].corr(eval_df[metric], method='spearman')
    kendall_corr = eval_df['manual_score'].corr(eval_df[metric], method='kendall')
    
    # Store results
    correlations[metric] = {
        'Spearman': spearman_corr,
        'Kendall': kendall_corr
    }

# Convert to DataFrame for better visualization
correlation_df = pd.DataFrame(correlations).T
print("Correlation between each metric and manual score:")
print(correlation_df)

Correlation between each metric and manual score:
             Spearman   Kendall
Exact Match       NaN       NaN
BLEU         0.613213  0.497908
ROUGE-L      0.612819  0.494840
METEOR       0.487628  0.379940
ChrF         0.424102  0.351727


/usr/lib/python3/dist-packages/scipy/stats/_stats_py.py:4529: SpearmanRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.
  warnings.warn(SpearmanRConstantInputWarning())


## Conclusion

The correlation results indicate that BLEU and ROUGE-L have the strongest relationships with the manual scores, as reflected in both Spearman and Kendall correlation coefficients (0.613 and 0.497 for BLEU, and 0.612 and 0.495 for ROUGE-L, respectively). This suggests that BLEU and ROUGE-L are more aligned with human judgment in evaluating the similarity of generated code completions, particularly in cases where exact structural and lexical match plays a role.